src/comm/Connection.cc

```diff
@@ -77,6 +79,21 @@ Comm::Connection::copyDetails() const
     return c;
 }
 
+Comm::ConnectionPointer
+Comm::Connection::cloneDestinationDetails() const
+{
+    const ConnectionPointer c = new Comm::Connection;
+    // Branch XXX: Do we really need to copy the local address? Does some peer
+    // selection code sets it? I doubt it is always set by the peer selection
+    // code. If some peer selection code paths set it and some do not, we have
+    // an even bigger problem because this copying would have to be conditional.
```

Comment:
>  If some peer selection code paths set it and some do not

Agreed: we need to copy the local address in the case of 'spoofing': peer selection code (PeerSelector::handlePath()) calls getOutgoingAddress(), which copies this address from request. I don't think we should re-calculate these getOutgoingAddress() conditions and pass the result into cloneDestinationDetails() - so that it could decide whether to copy the 'local' address or not. Much easier approach (IMO) would be to 'clone' the destination address each time it is requested from ResolvedPeers, thus guaranteeing that the connection fields are pristine. We would need to sacrifice our optimization with 'dirty' flag in this case. Do you agree?

Reply:
I do.